src/puppetlabs/puppetdb/http/command.clj

```diff
@@ -84,31 +84,30 @@
 
 (defn- blocking-submit-command
   "Submit a command by calling do-submit-fn and block until it completes.
-  Subscribes to response-pub on the topic of the commands uuid, waiting up to
-  completion-timeout-ms."
-  [do-submit-fn response-pub uuid completion-timeout-ms]
-  (with-chan [response-chan (async/chan)]
-    (with-sub {:from response-pub :topic (str uuid) :chan response-chan}
-      (let [timeout-chan (async/timeout completion-timeout-ms)
-            _ (do-submit-fn)]
-        (async/alt!!
-          timeout-chan (http/json-response {:uuid uuid
-                                            :processed false
-                                            :timed_out true}
-                                           http/status-unavailable)
-          response-chan ([{:keys [command exception]}]
-                         (let [base-response {:uuid uuid
-                                              :processed true}]
-                           (if exception
-                             (http/json-response
-                              (assoc base-response
-                                     :timed_out false
-                                     :error (str exception)
-                                     :stack_trace (map str (.getStackTrace exception)))
-                              http/status-unavailable)
-                             (http/json-response (assoc base-response
-                                                        :timed_out false)
-                                                 http/status-ok)))))))))
+  Waiting up to completion-timeout-ms if provided."
+  [do-submit-fn uuid completion-timeout-ms]
+  (let [command-result (promise)]
+    (do-submit-fn (fn [the-result]
+                    (deliver command-result the-result)))
```

Comment:
Previously this would watch a core.async/channel for results, this is now a callback with a promise. Note that this (like before) doesn't persist past restarts.